In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

## Kobe Bryant  shot selection

Cсылка на соревнование: https://www.kaggle.com/c/kobe-bryant-shot-selection

Goal: Fun and education

Using 20 years of data on Kobe's swishes and misses, can you predict which shots will find the bottom of the net? This competition is well suited for practicing classification basics, feature engineering, and time series analysis. Practice got Kobe an eight-figure contract and 5 championship rings. What will it get you?

This data contains the location and circumstances of every field goal attempted by Kobe Bryant took during his 20-year career. Your task is to predict whether the basket went in (shot_made_flag).

We have removed 5000 of the shot_made_flags (represented as missing values in the csv file). These are the test set shots for which you must submit a prediction. You are provided a sample submission file with the correct shot_ids needed for a valid prediction.

In [2]:
data = pd.read_csv('Kobe.csv')

In [3]:
data.head()

action_type combined_shot_type  game_event_id   game_id      lat  \
0          Jump Shot          Jump Shot             10  20000012  33.9723   
1          Jump Shot          Jump Shot             12  20000012  34.0443   
2          Jump Shot          Jump Shot             35  20000012  33.9093   
3          Jump Shot          Jump Shot             43  20000012  33.8693   
4  Driving Dunk Shot               Dunk            155  20000012  34.0443   

   loc_x  loc_y       lon  minutes_remaining  period   ...          shot_type  \
0    167     72 -118.1028                 10       1   ...     2PT Field Goal   
1   -157      0 -118.4268                 10       1   ...     2PT Field Goal   
2   -101    135 -118.3708                  7       1   ...     2PT Field Goal   
3    138    175 -118.1318                  6       1   ...     2PT Field Goal   
4      0      0 -118.2698                  6       2   ...     2PT Field Goal   

          shot_zone_area  shot_zone_basic  shot_zone_range     team_id  \
0          Right Side(R)        Mid-Range        16-24 ft.  1610612747   
1           Left Side(L)        Mid-Range         8-16 ft.  1610612747   
2   Left Side Center(LC)        Mid-Range        16-24 ft.  1610612747   
3  Right Side Center(RC)        Mid-Range        16-24 ft.  1610612747   
4              Center(C)  Restricted Area  Less Than 8 ft.  1610612747   

            team_name   game_date    matchup opponent  shot_id  
0  Los Angeles Lakers  2000-10-31  LAL @ POR      POR        1  
1  Los Angeles Lakers  2000-10-31  LAL @ POR      POR        2  
2  Los Angeles Lakers  2000-10-31  LAL @ POR      POR        3  
3  Los Angeles Lakers  2000-10-31  LAL @ POR      POR        4  
4  Los Angeles Lakers  2000-10-31  LAL @ POR      POR        5  

[5 rows x 25 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30697 entries, 0 to 30696
Data columns (total 25 columns):
action_type           30697 non-null object
combined_shot_type    30697 non-null object
game_event_id         30697 non-null int64
game_id               30697 non-null int64
lat                   30697 non-null float64
loc_x                 30697 non-null int64
loc_y                 30697 non-null int64
lon                   30697 non-null float64
minutes_remaining     30697 non-null int64
period                30697 non-null int64
playoffs              30697 non-null int64
season                30697 non-null object
seconds_remaining     30697 non-null int64
shot_distance         30697 non-null int64
shot_made_flag        25697 non-null float64
shot_type             30697 non-null object
shot_zone_area        30697 non-null object
shot_zone_basic       30697 non-null object
shot_zone_range       30697 non-null object
team_id               30697 non-null int64
team_name         

пропущенные значения есть только в таргете, которые мы должны предсказать

In [5]:
data['combined_shot_type'].value_counts()

Jump Shot    23485
Layup         5448
Dunk          1286
Tip Shot       184
Hook Shot      153
Bank Shot      141
Name: combined_shot_type, dtype: int64

In [6]:
data_for_an=data.dropna()

In [7]:
data_for_an.shape

(25697, 25)

 Описание данных
 Группа 1.Координаты
 -широта/долгота
 -loc_x/loc_y
 -shot_distance
 -shot_type
 Координаты географические и декартовы по сути содержат одну и ту же информацию-место точки удара на карте.
 Можно и нужно (для линейной регрессии) выбрать что-то одно, что по каким-то причинам будет работать лучше.
 shot distance и shot type то же одно и тоже- двух-или трехочковый бросок зависит от дальности.
 Информация об этом тоже содержится в первых двух переменных. Только еще в первых двух переменных содержится информация об угле.
 Группа 2.Переменные для группировки координат
 -shot_zone_area
 -shot_zone_basic
 -shot_zone_range
 В чем-то содержат пересекающуюся информацию, в чем-то разную. Характеризуют область на поле. Можно сгруппировать по ним координаты.
 Группа 3 Сезон
 - из даты игры вытянуть месяц - чтобы понять, если зависимость от части сезона (начало или конец)
 - season- года- посмотреть количество бросков на количество игр по сезонам.
 Группа 4 Время в игре
 Период+минуты (начало периода/середина/конец)
 Группа 3 (месяц) и 4 обработать как качественные переменные
 Группа 5. Тип броска
 action_type и combined_shot_type Лучше использовать второе (качественная переменная)

Проверим на выбросы и разделим на тест и трейн

In [8]:
data_for_an.describe(percentiles=(0.01, 0.99))

game_event_id       game_id           lat         loc_x         loc_y  \
count   25697.000000  2.569700e+04  25697.000000  25697.000000  25697.000000   
mean      249.348679  2.474109e+07     33.953043      7.148422     91.257345   
std       149.778520  7.738108e+06      0.088152    110.073147     88.152106   
min         2.000000  2.000001e+07     33.253300   -250.000000    -44.000000   
1%          5.000000  2.000021e+07     33.773300   -225.000000    -18.000000   
50%       253.000000  2.090034e+07     33.970300      0.000000     74.000000   
99%       545.000000  4.990003e+07     34.062300    228.000000    271.000000   
max       653.000000  4.990009e+07     34.088300    248.000000    791.000000   

                lon  minutes_remaining        period      playoffs  \
count  25697.000000       25697.000000  25697.000000  25697.000000   
mean    -118.262652           4.886796      2.520800      0.146243   
std        0.110073           3.452475      1.151626      0.353356   
min     -118.519800           0.000000      1.000000      0.000000   
1%      -118.494800           0.000000      1.000000      0.000000   
50%     -118.269800           5.000000      3.000000      0.000000   
99%     -118.041800          11.000000      5.000000      1.000000   
max     -118.021800          11.000000      7.000000      1.000000   

       seconds_remaining  shot_distance  shot_made_flag       team_id  \
count       25697.000000   25697.000000    25697.000000  2.569700e+04   
mean           28.311554      13.457096        0.446161  1.610613e+09   
std            17.523392       9.388725        0.497103  0.000000e+00   
min             0.000000       0.000000        0.000000  1.610613e+09   
1%              0.000000       0.000000        0.000000  1.610613e+09   
50%            28.000000      15.000000        0.000000  1.610613e+09   
99%            59.000000      28.000000        1.000000  1.610613e+09   
max            59.000000      79.000000        1.000000  1.610613e+09   

            shot_id  
count  25697.000000  
mean   15328.166946  
std     8860.462397  
min        2.000000  
1%       310.960000  
50%    15336.000000  
99%    30378.080000  
max    30697.000000

Из переменных для анализа loc_y имеет выброс и shot_distance

In [9]:
data_for_an=data_for_an[data_for_an['loc_y']<272]

In [10]:
data_for_an.describe(percentiles=(0.01, 0.99))

game_event_id       game_id           lat         loc_x         loc_y  \
count   25442.000000  2.544200e+04  25442.000000  25442.000000  25442.000000   
mean      248.897257  2.473374e+07     33.955879      7.236459     88.421468   
std       149.776246  7.731662e+06      0.082943    110.138580     82.943264   
min         2.000000  2.000001e+07     33.773300   -250.000000    -44.000000   
1%          5.000000  2.000021e+07     33.785300   -225.000000    -18.000000   
50%       253.000000  2.090033e+07     33.973300      0.000000     71.000000   
99%       544.000000  4.990003e+07     34.062300    228.000000    259.000000   
max       653.000000  4.990009e+07     34.088300    248.000000    271.000000   

                lon  minutes_remaining        period      playoffs  \
count  25442.000000       25442.000000  25442.000000  25442.000000   
mean    -118.262564           4.921193      2.520439      0.146136   
std        0.110139           3.441681      1.151813      0.353250   
min     -118.519800           0.000000      1.000000      0.000000   
1%      -118.494800           0.000000      1.000000      0.000000   
50%     -118.269800           5.000000      3.000000      0.000000   
99%     -118.041800          11.000000      5.000000      1.000000   
max     -118.021800          11.000000      7.000000      1.000000   

       seconds_remaining  shot_distance  shot_made_flag       team_id  \
count       25442.000000   25442.000000    25442.000000  2.544200e+04   
mean           28.484985      13.208474        0.449139  1.610613e+09   
std            17.447462       9.009959        0.497416  0.000000e+00   
min             0.000000       0.000000        0.000000  1.610613e+09   
1%              0.000000       0.000000        0.000000  1.610613e+09   
50%            28.000000      15.000000        0.000000  1.610613e+09   
99%            59.000000      27.000000        1.000000  1.610613e+09   
max            59.000000      31.000000        1.000000  1.610613e+09   

            shot_id  
count  25442.000000  
mean   15321.976889  
std     8855.274562  
min        2.000000  
1%       311.410000  
50%    15318.500000  
99%    30376.590000  
max    30697.000000

In [11]:
from sklearn.model_selection import train_test_split

data_train, data_test = train_test_split(data_for_an, test_size=0.2)
len(data_train)

20353

In [12]:
groups_coordinates=data_train.groupby(['shot_zone_area','shot_zone_basic','shot_zone_range']).mean()[['loc_x','loc_y']]

In [13]:
groups_coordinates

loc_x  \
shot_zone_area        shot_zone_basic       shot_zone_range               
Center(C)             Above the Break 3     24+ ft.            1.705344   
                      In The Paint (Non-RA) 8-16 ft.           3.323108   
                                            Less Than 8 ft.    2.567412   
                      Mid-Range             16-24 ft.          0.785539   
                                            8-16 ft.          -1.250000   
                      Restricted Area       Less Than 8 ft.    0.307301   
Left Side Center(LC)  Above the Break 3     24+ ft.         -161.105769   
                      Mid-Range             16-24 ft.       -114.474110   
Left Side(L)          In The Paint (Non-RA) 8-16 ft.         -66.211382   
                      Left Corner 3         24+ ft.         -232.828877   
                      Mid-Range             16-24 ft.       -174.275572   
                                            8-16 ft.        -118.458949   
Right Side Center(RC) Above the Break 3     24+ ft.          155.864759   
                      Mid-Range             16-24 ft.        113.427774   
Right Side(R)         In The Paint (Non-RA) 8-16 ft.          68.593472   
                      Mid-Range             16-24 ft.        173.509862   
                                            8-16 ft.         119.667579   
                      Right Corner 3        24+ ft.          231.516484   

                                                                  loc_y  
shot_zone_area        shot_zone_basic       shot_zone_range              
Center(C)             Above the Break 3     24+ ft.          252.377099  
                      In The Paint (Non-RA) 8-16 ft.         108.111452  
                                            Less Than 8 ft.   43.540575  
                      Mid-Range             16-24 ft.        190.776961  
                                            8-16 ft.         148.129630  
                      Restricted Area       Less Than 8 ft.    3.086163  
Left Side Center(LC)  Above the Break 3     24+ ft.          196.496566  
                      Mid-Range             16-24 ft.        151.288026  
Left Side(L)          In The Paint (Non-RA) 8-16 ft.          76.256098  
                      Left Corner 3         24+ ft.           18.711230  
                      Mid-Range             16-24 ft.         50.264741  
                                            8-16 ft.          40.713465  
Right Side Center(RC) Above the Break 3     24+ ft.          197.698960  
                      Mid-Range             16-24 ft.        152.897086  
Right Side(R)         In The Paint (Non-RA) 8-16 ft.          75.899110  
                      Mid-Range             16-24 ft.         42.452663  
                                            8-16 ft.          35.364569  
                      Right Corner 3        24+ ft.           18.157509

In [14]:
data_train = pd.merge(data_train, groups_coordinates,  how='left', left_on=['shot_zone_area', 'shot_zone_basic', 'shot_zone_range'], right_on = ['shot_zone_area', 'shot_zone_basic', 'shot_zone_range'])

In [15]:
data_test = pd.merge(data_test, groups_coordinates,  how='left', left_on=['shot_zone_area', 'shot_zone_basic', 'shot_zone_range'], right_on = ['shot_zone_area', 'shot_zone_basic', 'shot_zone_range'])

In [16]:
from datetime import datetime
data_1=[]
for i in data_train['game_date'].values:
    a=datetime.strptime(i, '%Y-%m-%d')
    a=a.month
    data_1.append(a)
data_train['month']=data_1
data_2=[]
for i in data_test['game_date'].values:
    a=datetime.strptime(i, '%Y-%m-%d')
    a=a.month
    data_2.append(a)
data_test['month']=data_2

In [17]:
data_season=data_train.groupby(['season','game_id']).count()

In [18]:
data_season=data_season.groupby('season').count()[['action_type','lat']]

In [19]:
data_season_1=data_train.groupby(['season']).count()[['action_type','lat']]

In [20]:
data_season=data_season.merge(data_season_1, how='left', left_on='season', right_on='season')
data_season['shot_per_game']=data_season['lat_y']/data_season['lat_x']
data_season['shots']=data_season['lat_y']
data_season=data_season[['shot_per_game','shots']]

In [21]:
data_train = pd.merge(data_train, data_season,  how='left', left_on=['season'], right_on = ['season'])
data_test = pd.merge(data_test, data_season,  how='left', left_on=['season'], right_on = ['season'])

In [22]:
def minutes_category(db):
    minutes_cat=[]
    for i in db['minutes_remaining'].values:
        if i<=4:
            minutes_cat.append('finish')
        else:
            if i<=8:
                minutes_cat.append('middle')  
            else:
                minutes_cat.append('start')
    print (len(minutes_cat))
    db['minutes_cat']=minutes_cat

In [23]:
minutes_category(data_train)
minutes_category(data_test)

20353
5089


In [24]:
period_str=[]
for i in data_train['period']:
    i=str(i)
    period_str.append(i)

In [25]:
data_train['period_str']=period_str
data_train['period_min']=data_train['period_str']+data_train['minutes_cat']

In [26]:
period_str=[]
for i in data_test['period']:
    i=str(i)
    period_str.append(i)
data_test['period_str']=period_str
data_test['period_min']=data_test['period_str']+data_test['minutes_cat']

In [27]:
data_train['locxLocy']=data_train['loc_x_y']*data_train['loc_y_y']
data_test['locxLocy']=data_test['loc_x_y']*data_test['loc_y_y']

In [28]:
data_train_for_an=data_train[['combined_shot_type','loc_x_y','loc_y_y','locxLocy','month','shot_per_game', 'period_min','shot_made_flag']]

In [29]:
data_test_for_an=data_test[['combined_shot_type','loc_x_y','loc_y_y','locxLocy','month','shot_per_game', 'period_min','shot_made_flag']]

In [30]:
def get_woe_v1(db, db1, col, target_col):
    all_good = len(db[db[target_col] == 1][col])
    all_bad = len(db[db[target_col] == 0][col])
    odds_series = (
        db[db[target_col] == 1][col].value_counts()
        /
        db[db[target_col] == 0][col].value_counts()
    )
    odds_series = odds_series / all_good * all_bad
    category_woe_dict = np.log(odds_series).to_dict()
    db[col + '_woe'] = db[col].apply(category_woe_dict.get)
    db1[col + '_woe'] = db1[col].apply(category_woe_dict.get)
    return db, db1

In [31]:
res=['combined_shot_type', 'month', 'period_min']
for i in res:
    get_woe_v1(data_train_for_an, data_test_for_an, i, 'shot_made_flag')


C:\Users\volko\Anaconda3\lib\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # This is added back by InteractiveShellApp.init_path()
C:\Users\volko\Anaconda3\lib\site-packages\ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if sys.path[0] == '':


In [32]:
data_train_for_an=data_train_for_an.drop(res, axis=1)
data_test_for_an=data_test_for_an.drop(res, axis=1)

In [33]:
corr_matrix = data_train_for_an.corr()
corr_matrix

loc_x_y   loc_y_y  locxLocy  shot_per_game  \
loc_x_y                 1.000000 -0.016185  0.824248       0.018836   
loc_y_y                -0.016185  1.000000  0.025392       0.030795   
locxLocy                0.824248  0.025392  1.000000       0.013693   
shot_per_game           0.018836  0.030795  0.013693       1.000000   
shot_made_flag          0.003458 -0.146490  0.009800       0.016991   
combined_shot_type_woe -0.027488 -0.429835 -0.015314      -0.037157   
month_woe              -0.005963 -0.000955 -0.010171       0.030856   
period_min_woe          0.003836 -0.057119  0.002461       0.006037   

                        shot_made_flag  combined_shot_type_woe  month_woe  \
loc_x_y                       0.003458               -0.027488  -0.005963   
loc_y_y                      -0.146490               -0.429835  -0.000955   
locxLocy                      0.009800               -0.015314  -0.010171   
shot_per_game                 0.016991               -0.037157   0.030856   
shot_made_flag                1.000000                0.243080   0.015777   
combined_shot_type_woe        0.243080                1.000000   0.001463   
month_woe                     0.015777                0.001463   1.000000   
period_min_woe                0.057896                0.007232  -0.009772   

                        period_min_woe  
loc_x_y                       0.003836  
loc_y_y                      -0.057119  
locxLocy                      0.002461  
shot_per_game                 0.006037  
shot_made_flag                0.057896  
combined_shot_type_woe        0.007232  
month_woe                    -0.009772  
period_min_woe                1.000000

In [34]:
from sklearn.linear_model import Lasso
from sklearn.metrics import roc_auc_score

def lasso_regression(table, predictors, alpha):
    lassoreg = Lasso(alpha=alpha, normalize=True, max_iter=1e5)
    lassoreg.fit(predictors, table['shot_made_flag'])
    y_pred = lassoreg.predict(predictors)
    roc_auc = np.round(roc_auc_score(table['shot_made_flag'], y_pred), 2)
    rss = sum((y_pred-table['shot_made_flag']) ** 2)
    ret = [rss, roc_auc]
    ret.extend([lassoreg.intercept_])
    ret.extend(lassoreg.coef_)
    return ret

In [35]:
predictors=data_train_for_an.drop('shot_made_flag', axis=1)
predictors1=data_test_for_an.drop('shot_made_flag', axis=1)

In [36]:
alpha_lasso = [1e-15, 1e-10, 1e-8, 1e-5,1e-4, 1e-3,1e-2, 1, 5, 10]

# create dataframe to store coefficients
col = ['rss','roc_auc', 'intercept'] + predictors.columns.tolist()
ind = ['alpha_%.2g' % alpha_lasso[i] for i in range(0, 10)]
coef_matrix_lasso = pd.DataFrame(index=ind, columns=col)

#Iterate over the 10 alpha values:
for i in range(10):
    coef_matrix_lasso.iloc[i, ] = lasso_regression(data_train_for_an, predictors, alpha_lasso[i])

In [37]:
coef_matrix_lasso

rss roc_auc intercept      loc_x_y      loc_y_y  \
alpha_1e-15   4706.02    0.62  0.397242   -5.042e-05 -0.000314066   
alpha_1e-10   4706.02    0.62  0.397242 -5.04193e-05 -0.000314066   
alpha_1e-08   4706.02    0.62  0.397248 -5.03419e-05 -0.000314048   
alpha_1e-05   4706.43    0.62  0.404136           -0 -0.000298034   
alpha_0.0001  4721.46    0.62  0.460635            0 -0.000189265   
alpha_0.001   5034.65     0.5  0.448681            0           -0   
alpha_0.01    5034.65     0.5  0.448681            0           -0   
alpha_1       5034.65     0.5  0.448681            0           -0   
alpha_5       5034.65     0.5  0.448681            0           -0   
alpha_10      5034.65     0.5  0.448681            0           -0   

                 locxLocy shot_per_game combined_shot_type_woe month_woe  \
alpha_1e-15   8.21326e-07     0.0053768               0.174264   0.23971   
alpha_1e-10    8.2132e-07     0.0053768               0.174264   0.23971   
alpha_1e-08   8.20728e-07    0.00537618               0.174263  0.239665   
alpha_1e-05   3.98544e-07    0.00476403               0.172889  0.194969   
alpha_0.0001            0             0               0.157775         0   
alpha_0.001             0             0                      0         0   
alpha_0.01              0             0                      0         0   
alpha_1                 0             0                      0         0   
alpha_5                 0             0                      0         0   
alpha_10                0             0                      0         0   

             period_min_woe  
alpha_1e-15        0.225763  
alpha_1e-10        0.225762  
alpha_1e-08        0.225751  
alpha_1e-05        0.214213  
alpha_0.0001       0.110064  
alpha_0.001               0  
alpha_0.01                0  
alpha_1                   0  
alpha_5                   0  
alpha_10                  0

Обнуляются месяц, среднее количество ударов за игру, координата X и совокупная координата

In [38]:
# create dataframe to store coefficients
col = ['rss','roc_auc', 'intercept'] + predictors.columns.tolist()
ind = ['alpha_%.2g' % alpha_lasso[i] for i in range(0, 10)]
coef_matrix_lasso_test = pd.DataFrame(index=ind, columns=col)

#Iterate over the 10 alpha values:
for i in range(10):
    coef_matrix_lasso_test.iloc[i, ] = lasso_regression(data_test_for_an, predictors1, alpha_lasso[i])
coef_matrix_lasso_test

rss roc_auc intercept      loc_x_y      loc_y_y  \
alpha_1e-15   1185.33     0.6  0.433263 -7.91493e-05 -0.000105024   
alpha_1e-10   1185.33     0.6  0.433263 -7.91489e-05 -0.000105024   
alpha_1e-08   1185.33     0.6  0.433266 -7.91112e-05 -0.000105015   
alpha_1e-05   1185.37     0.6  0.436933 -4.14814e-05 -9.58405e-05   
alpha_0.0001  1186.26     0.6  0.448024           -0 -3.99582e-05   
alpha_0.001   1212.25    0.59  0.448144           -0           -0   
alpha_0.01    1260.02     0.5  0.450973           -0           -0   
alpha_1       1260.02     0.5  0.450973           -0           -0   
alpha_5       1260.02     0.5  0.450973           -0           -0   
alpha_10      1260.02     0.5  0.450973           -0           -0   

                 locxLocy shot_per_game combined_shot_type_woe month_woe  \
alpha_1e-15   5.16213e-07    0.00149895               0.183857  0.173264   
alpha_1e-10    5.1621e-07    0.00149895               0.183857  0.173264   
alpha_1e-08   5.15917e-07    0.00149862               0.183856  0.173242   
alpha_1e-05   2.22898e-07    0.00116792               0.183198  0.151172   
alpha_0.0001           -0             0               0.175926         0   
alpha_0.001            -0            -0              0.0770515         0   
alpha_0.01             -0            -0                      0         0   
alpha_1                -0            -0                      0         0   
alpha_5                -0            -0                      0         0   
alpha_10               -0            -0                      0         0   

             period_min_woe  
alpha_1e-15       0.0736879  
alpha_1e-10       0.0736879  
alpha_1e-08       0.0736826  
alpha_1e-05        0.068328  
alpha_0.0001      0.0172615  
alpha_0.001               0  
alpha_0.01                0  
alpha_1                   0  
alpha_5                   0  
alpha_10                  0

In [39]:
from sklearn.linear_model import Ridge

def ridge_regression(data, predictors, alpha):
    ridgereg = Ridge(alpha=alpha, normalize=True)
    ridgereg.fit(predictors, data['shot_made_flag'])
    y_pred = ridgereg.predict(predictors)
    rss = sum((y_pred-data['shot_made_flag']) ** 2)
    roc_auc = np.round(roc_auc_score(data['shot_made_flag'], y_pred), 2)
    ret = [rss, roc_auc]
    ret.extend([ridgereg.intercept_])
    ret.extend(ridgereg.coef_)
    return ret

In [40]:
alpha_ridge = [1e-15, 1e-10, 1e-8, 1e-4, 1e-3, 1e-2, 1, 5, 10, 20]
col = ['rss','roc_auc', 'intercept'] + predictors.columns.tolist()
ind = ['alpha_%.2g' % alpha_ridge[i] for i in range(0, 10)]
coef_matrix_ridge = pd.DataFrame(index=ind, columns=col)

for i in range(10):
    coef_matrix_ridge.iloc[i, ] = ridge_regression(data_train_for_an, predictors, alpha_ridge[i])

In [41]:
coef_matrix_ridge

rss roc_auc intercept      loc_x_y      loc_y_y  \
alpha_1e-15   4706.02    0.62  0.397242   -5.042e-05 -0.000314066   
alpha_1e-10   4706.02    0.62  0.397242   -5.042e-05 -0.000314066   
alpha_1e-08   4706.02    0.62  0.397242   -5.042e-05 -0.000314066   
alpha_0.0001  4706.02    0.62  0.397255 -5.03845e-05 -0.000314101   
alpha_0.001   4706.02    0.62  0.397371 -5.00661e-05 -0.000314416   
alpha_0.01    4706.05    0.62  0.398517  -4.7051e-05 -0.000317465   
alpha_1       4773.69    0.62  0.441834  7.42003e-07 -0.000310648   
alpha_5       4915.05    0.62  0.450852  2.05845e-06  -0.00013717   
alpha_10      4964.22    0.62  0.450469  1.28835e-06  -7.9241e-05   
alpha_20      4996.08    0.61  0.449804  7.20216e-07  -4.2865e-05   

                 locxLocy shot_per_game combined_shot_type_woe  month_woe  \
alpha_1e-15   8.21326e-07     0.0053768               0.174264    0.23971   
alpha_1e-10   8.21326e-07     0.0053768               0.174264    0.23971   
alpha_1e-08   8.21326e-07     0.0053768               0.174264    0.23971   
alpha_0.0001  8.21013e-07    0.00537612               0.174245   0.239687   
alpha_0.001   8.18207e-07       0.00537               0.174071   0.239476   
alpha_0.01    7.91471e-07    0.00530953                0.17236   0.237392   
alpha_1       2.22127e-07    0.00230006              0.0871032   0.121873   
alpha_5       6.25738e-08   0.000655825              0.0305355  0.0410383   
alpha_10      3.30244e-08    0.00034168              0.0169314  0.0224275   
alpha_20      1.69758e-08   0.000173929             0.00895969  0.0117597   

             period_min_woe  
alpha_1e-15        0.225763  
alpha_1e-10        0.225763  
alpha_1e-08        0.225763  
alpha_0.0001       0.225739  
alpha_0.001        0.225531  
alpha_0.01         0.223472  
alpha_1            0.114678  
alpha_5           0.0396949  
alpha_10          0.0218998  
alpha_20          0.0115517

минимальное количество переменных без снижения качества достигается при alpha_1. Сильнее всего близки к 0 координата x, скомбинированная x-y

In [42]:
alpha_ridge = [1e-15, 1e-10, 1e-8, 1e-4, 1e-3, 1e-2, 1, 5, 10, 20]
col = ['rss','roc_auc', 'intercept'] + predictors.columns.tolist()
ind = ['alpha_%.2g' % alpha_ridge[i] for i in range(0, 10)]
coef_matrix_ridge1 = pd.DataFrame(index=ind, columns=col)

for i in range(10):
    coef_matrix_ridge1.iloc[i, ] = ridge_regression(data_test_for_an, predictors1, alpha_ridge[i])

In [43]:
coef_matrix_ridge1

rss roc_auc intercept      loc_x_y      loc_y_y  \
alpha_1e-15   1185.33     0.6  0.433263 -7.91493e-05 -0.000105024   
alpha_1e-10   1185.33     0.6  0.433263 -7.91493e-05 -0.000105024   
alpha_1e-08   1185.33     0.6  0.433263 -7.91493e-05 -0.000105024   
alpha_0.0001  1185.33     0.6  0.433275 -7.91181e-05  -0.00010509   
alpha_0.001   1185.33     0.6  0.433388 -7.88393e-05 -0.000105686   
alpha_0.01    1185.34     0.6  0.434493 -7.61837e-05 -0.000111495   
alpha_1       1202.13     0.6  0.467118 -2.28906e-05 -0.000221428   
alpha_5       1233.93    0.59  0.462032 -8.32665e-06  -0.00010863   
alpha_10       1244.7    0.59  0.457769 -4.80237e-06 -6.37483e-05   
alpha_20      1251.64    0.59  0.454769 -2.61198e-06 -3.47673e-05   

                 locxLocy shot_per_game combined_shot_type_woe   month_woe  \
alpha_1e-15   5.16213e-07    0.00149895               0.183857    0.173264   
alpha_1e-10   5.16213e-07    0.00149895               0.183857    0.173264   
alpha_1e-08   5.16213e-07    0.00149895               0.183857    0.173264   
alpha_0.0001  5.15961e-07    0.00149851               0.183835    0.173251   
alpha_0.001   5.13702e-07    0.00149454               0.183637    0.173135   
alpha_0.01    4.92159e-07    0.00145547                0.18169    0.171974   
alpha_1       4.93146e-08   8.43802e-06              0.0887054   0.0924282   
alpha_5      -1.38169e-08  -0.000190636              0.0305679    0.030648   
alpha_10     -1.19529e-08  -0.000131305              0.0168932   0.0166203   
alpha_20     -7.68928e-09  -7.73319e-05             0.00892337  0.00867012   

             period_min_woe  
alpha_1e-15       0.0736879  
alpha_1e-10       0.0736879  
alpha_1e-08       0.0736879  
alpha_0.0001      0.0736781  
alpha_0.001       0.0735901  
alpha_0.01        0.0727232  
alpha_1           0.0348344  
alpha_5           0.0124457  
alpha_10         0.00696454  
alpha_20          0.0037074

In [44]:
from sklearn.linear_model import LinearRegression

lin_reg = LinearRegression()
lin_reg.fit(predictors, data_train_for_an['shot_made_flag'])
y_pred = lin_reg.predict(predictors)
roc_auc = np.round(roc_auc_score(data_train_for_an['shot_made_flag'], y_pred), 2)
rss = sum((y_pred-data_train_for_an['shot_made_flag']) ** 2)
print (rss, roc_auc)

4706.0227099757585 0.62


In [45]:
lin_reg1 = LinearRegression()
lin_reg1.fit(predictors1, data_test_for_an['shot_made_flag'])
y_pred1 = lin_reg1.predict(predictors1)
roc_auc = np.round(roc_auc_score(data_test_for_an['shot_made_flag'], y_pred1), 2)
rss = sum((y_pred1-data_test_for_an['shot_made_flag']) ** 2)
print (rss, roc_auc)

1185.3271216693693 0.6


In [46]:
data_train.columns

Index(['action_type', 'combined_shot_type', 'game_event_id', 'game_id', 'lat',
       'loc_x_x', 'loc_y_x', 'lon', 'minutes_remaining', 'period', 'playoffs',
       'season', 'seconds_remaining', 'shot_distance', 'shot_made_flag',
       'shot_type', 'shot_zone_area', 'shot_zone_basic', 'shot_zone_range',
       'team_id', 'team_name', 'game_date', 'matchup', 'opponent', 'shot_id',
       'loc_x_y', 'loc_y_y', 'month', 'shot_per_game', 'shots', 'minutes_cat',
       'period_str', 'period_min', 'locxLocy'],
      dtype='object')

In [47]:

data_train_fac=data_train.drop(['action_type','period', 'game_event_id', 'game_id','season','minutes_remaining','shot_type', 'shot_zone_area', 'shot_zone_basic', 'shot_zone_range',
'team_id', 'team_name',  'matchup', 'shot_id','shot_per_game',  'game_date','minutes_cat','period_str'], axis=1)
data_test_fac=data_test.drop(['action_type','period', 'game_event_id', 'game_id','minutes_remaining','season','shot_type', 'shot_zone_area', 'shot_zone_basic', 'shot_zone_range',
       'team_id', 'team_name',  'matchup', 'shot_id',
        'shot_per_game',  'game_date','minutes_cat',
       'period_str'], axis=1)
res=['combined_shot_type', 'month', 'period_min', 'opponent']
for i in res:
    get_woe_v1(data_train_fac, data_test_fac, i, 'shot_made_flag')
predictors2=data_train_fac.drop('shot_made_flag', axis=1)
predictors2=predictors2.drop(res,axis=1)
from sklearn.decomposition import PCA
pca = PCA(n_components=predictors2.shape[1]).fit(predictors2)
for i in range(1, predictors2.shape[1] + 1):
    print("n-components=%s, explained variance %.3f" % (i, sum(pca.explained_variance_ratio_[:i])))

n-components=1, explained variance 0.999
n-components=2, explained variance 1.000
n-components=3, explained variance 1.000
n-components=4, explained variance 1.000
n-components=5, explained variance 1.000
n-components=6, explained variance 1.000
n-components=7, explained variance 1.000
n-components=8, explained variance 1.000
n-components=9, explained variance 1.000
n-components=10, explained variance 1.000
n-components=11, explained variance 1.000
n-components=12, explained variance 1.000
n-components=13, explained variance 1.000
n-components=14, explained variance 1.000
n-components=15, explained variance 1.000


In [48]:
new_factors = PCA(n_components=2).fit_transform(predictors2)
model2 = LinearRegression().fit(new_factors, data_train_fac['shot_made_flag'])
model2.score(new_factors, data_train_fac['shot_made_flag'])

0.0010050732145289176

плохо описывают

следующий этап - проанализировать решения участников на кагле. Больше всего понра решение с площадями.
https://www.kaggle.com/selfishgene/psychology-of-a-professional-athlete
Хоть там и подход несколько иной, и информации о качестве я не нашла, но истина где-то в этом
Качество других решений варьируется от 0.6 до 0.7, те манипуляции с данными не приносят существенных результатов.
Каждый делает по -своему. а качесвто одно и то же.

В том примере исследователь считает площади и процент успешных попыток.
По аналогии, но более тупо- просто посчитаем WOE по сгруппир критеряям
shot_zone_area        30697 non-null object
shot_zone_basic       30697 non-null object
shot_zone_range       30697 non-null object

In [49]:
data_train['combined']=data_train['shot_zone_area']+data_train['shot_zone_basic']+data_train['shot_zone_range']
data_test['combined']=data_test['shot_zone_area']+data_test['shot_zone_basic']+data_test['shot_zone_range']

In [51]:
columns_2=['combined_shot_type', 'month', 'period_min','shot_per_game', 'combined', 'shot_made_flag']
data_tr_an2=data_train[columns_2]
data_test_an2=data_test[columns_2]

In [52]:
columns_woe=['combined_shot_type', 'month', 'period_min', 'combined']
for i in columns_woe:
    get_woe_v1(data_tr_an2, data_test_an2, i, 'shot_made_flag')

C:\Users\volko\Anaconda3\lib\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # This is added back by InteractiveShellApp.init_path()
C:\Users\volko\Anaconda3\lib\site-packages\ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if sys.path[0] == '':


In [53]:
data_tr_an2=data_tr_an2.drop(columns_woe, axis=1)
data_test_an2=data_test_an2.drop(columns_woe, axis=1)

In [54]:
predictors_w=data_tr_an2.drop('shot_made_flag', axis=1)
predictors_w1=data_test_an2.drop('shot_made_flag', axis=1)

In [55]:
lin_reg_w = LinearRegression()
lin_reg_w.fit(predictors_w, data_tr_an2['shot_made_flag'])
y_pred_w = lin_reg_w.predict(predictors_w)
roc_auc = np.round(roc_auc_score(data_tr_an2['shot_made_flag'], y_pred_w), 2)
rss = sum((y_pred_w-data_tr_an2['shot_made_flag']) ** 2)
print (rss, roc_auc)

4694.960270715225 0.63


можно сказать, что не лучше. Можно было б, как в примере- выявить типы стратегий игрокам по сегментам поля и ситуациям.
Возможно, это дало бы результат.

**Задания:**

1. Провести анализ данных. Много хороших примеров анализа можно посмотреть здесь https://www.kaggle.com/c/kobe-bryant-shot-selection/kernels
2. Подготовить фичи для обучения модели - нагенерить признаков, обработать пропущенные значения, проверить на возможные выбросы, обработать категориальные признаки и др.
3. Обучить линейную модель, Lasso, Ridge на тех же признаках - построить сравнительную таблицу коэффициентов, сделать заключения о том, как меняется величина коэффициентов, какие зануляются. Посчитать RSS

**Дополнительно**
4. Сравнить результаты на тестовом наборе данных - сделать train_test_split в самом начале, подготовить переменные, сравнить результаты работы классификаторов (те же 3), метрика ROC AUC
5. Построить PCA на подготовленных признаках, посмотреть, какие компоненты составляют наибольшую часть дисперсии целевой переменной